File: QA_Interview_Questions/04-array-deduplication.ts
/**
 * Question 4: Array Deduplication
 * Problem: Remove duplicate elements from an array while preserving order
 * 
 * Example:
 * Input: [1, 2, 2, 3, 3, 3, 4]
 * Output: [1, 2, 3, 4]
 */

function removeDuplicates<T>(arr: T[]): T[] {
  return [...new Set(arr)];
}

// Alternative: Using filter
function removeDuplicatesFilter<T>(arr: T[]): T[] {
  return arr.filter((value, index, self) => self.indexOf(value) === index);
}

// For objects/complex types
function removeDuplicatesWithKey<T>(arr: T[], key: keyof T): T[] {
  const seen = new Set<T[keyof T]>();
  return arr.filter((item) => {
    if (seen.has(item[key])) {
      return false;
    }
    seen.add(item[key]);
    return true;
  });
}

// Test cases
console.log(removeDuplicates([1, 2, 2, 3, 3, 3, 4])); // [1, 2, 3, 4]
console.log(removeDuplicates(["a", "b", "b", "c"])); // ['a', 'b', 'c']
console.log(removeDuplicatesFilter([5, 5, 10, 10, 15])); // [5, 10, 15]

interface User {
  id: number;
  name: string;
}

const users: User[] = [
  { id: 1, name: "Alice" },
  { id: 2, name: "Bob" },
  { id: 1, name: "Alice" }
];

console.log(removeDuplicatesWithKey(users, "id"));
// [{ id: 1, name: 'Alice' }, { id: 2, name: 'Bob' }]

File: Java_Interview_Questions/04-ArrayDeduplication.java
/**
 * Question 4: Array Deduplication
 * Problem: Remove duplicate elements from an array while preserving order
 * 
 * Example:
 * Input: [1, 2, 2, 3, 3, 3, 4]
 * Output: [1, 2, 3, 4]
 */

import java.util.*;

public class ArrayDeduplication {
    
    // Using HashSet
    public static int[] removeDuplicates(int[] arr) {
        Set<Integer> seen = new LinkedHashSet<>();
        for (int num : arr) {
            seen.add(num);
        }
        return seen.stream().mapToInt(Integer::intValue).toArray();
    }
    
    // Using HashSet for objects
    public static <T> List<T> removeDuplicatesGeneric(List<T> list) {
        Set<T> seen = new LinkedHashSet<>(list);
        return new ArrayList<>(seen);
    }
    
    // For sorted array - two pointer approach
    public static int removeDuplicatesSorted(int[] nums) {
        if (nums.length == 0) return 0;
        
        int i = 0;
        for (int j = 1; j < nums.length; j++) {
            if (nums[j] != nums[i]) {
                i++;
                nums[i] = nums[j];
            }
        }
        return i + 1;
    }
    
    // Remove duplicates with custom key
    public static <T> List<T> removeDuplicatesWithKey(List<T> list, 
                                                       java.util.function.Function<T, ?> keyExtractor) {
        Map<Object, T> seen = new LinkedHashMap<>();
        for (T item : list) {
            seen.putIfAbsent(keyExtractor.apply(item), item);
        }
        return new ArrayList<>(seen.values());
    }
    
    public static void main(String[] args) {
        System.out.println(Arrays.toString(removeDuplicates(new int[]{1, 2, 2, 3, 3, 3, 4}))); 
        // [1, 2, 3, 4]
        
        List<String> list = Arrays.asList("a", "b", "b", "c");
        System.out.println(removeDuplicatesGeneric(list)); 
        // [a, b, c]
        
        int[] arr = {0, 0, 1, 1, 1, 2, 2, 3, 3, 4};
        int len = removeDuplicatesSorted(arr);
        System.out.println(len); // 5
    }
}
